pkg/sign/manifest.go

```diff
@@ -0,0 +1,207 @@
+)
+
+// CertManagerManifests takes a path to a cert-manager-manifests.tar.gz file, loads it into
+// memory and signs anything inside the archive which is signable; currently,
+// the helm chart located at "deploy/chart/cert-manager.tgz" is signed, and a
+// signature "deploy/chart/cert-manager.tgz.prov" will be added.
+// The cert-manifests.tar.gz file is changed in-place.
+func CertManagerManifests(ctx context.Context, key string, path string) error {
+	if key == "" {
+		return fmt.Errorf("can't sign manifests without a signing key specified")
+	}
+
+	manifestLocation := "deploy/chart/cert-manager.tgz"
+
+	// 1. Create temp dir for chart archive to be extracted to
+	// (Helm signing requires a filename, not a reader, so we have to write to disk here)
+	tmpDest, err := os.MkdirTemp("", "extracted-manifests-")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary directory for extracting manifests: %w", err)
+	}
+
+	defer func() {
+		err := os.RemoveAll(tmpDest)
+		if err != nil {
+			log.Printf("failed to remove temporary directory %q: %v", tmpDest, err)
+		}
+	}()
+
+	// 2. Read + ungzip the manifests archive into memory
+	tarData, originalMode, err := ungzipManifestArchive(path)
+	if err != nil {
+		return err
+	}
+
+	chartFileData, err := tarball.ReadSingleFile(manifestLocation, bytes.NewReader(tarData))
+	if err != nil {
+		return fmt.Errorf("failed to read %q from %q: %w", manifestLocation, path, err)
+	}
+
+	// Write file to temp location (for helm's sake)
+	chartPath := filepath.Join(tmpDest, "cert-manager.tgz")
+	err = os.WriteFile(chartPath, chartFileData, 0o644)
+	if err != nil {
+		return err
+	}
+
+	// 3. Sign chart
+	signatureBytes, err := HelmChart(ctx, key, chartPath)
+	if err != nil {
+		return fmt.Errorf("failed to sign helm chart at %q: %w", chartPath, err)
+	}
+
+	// 4. Chmod the archive if needed so that it's writable, with a defer to reset its permissions
+	// after we're done. This is required because bazel forces the mode to 0o555.
+	modeResetFunc, err := ensureWritable(path, originalMode)
+	if err != nil {
+		return err
+	}
+
+	defer modeResetFunc()
+
+	// 5. Append file to original tar archive
+
+	// The tar spec requires that the end include two full empty blocks, so any valid tar file
+	// will have exactly 512 * 2 bytes of empty space at the end.
+	// We copy until the beginning of these empty blocks, then add our new header and our file,
+	// and then close the tar.Writer to complete the tar archive.
+	// See https://stackoverflow.com/a/18330903/1615417 for more details
+
+	// NB: We can't just open the file as O_APPEND and seek back 1024 bytes because it's gzipped
+	provPath := manifestLocation + ".prov"
+
+	newTar, err := signatureToTar(signatureBytes, provPath, 0o644)
+	if err != nil {
+		return err
+	}
+
+	targzOut := &bytes.Buffer{}
+	gzipWriter := gzip.NewWriter(targzOut)
+
+	_, err = gzipWriter.Write(append(tarData[:len(tarData)-1024], newTar...))
```

Comment:
My fear here is that this seems quite fragile for limited gain trying to be "smart". The [GNU website](https://www.gnu.org/software/tar/manual/html_node/Standard.html) itself states that while *reasonable* systems should add this two block trailing zeros, tar itself should handle when it does not.

"
At the end of the archive file there are two 512-byte blocks filled with binary zeros as an end-of-file marker. A reasonable system should write such end-of-file marker at the end of an archive, but must not assume that such a block exists when reading an archive. In particular GNU tar always issues a warning if it does not encounter it.
"

Since we are dealing with Kb rather than Tb, I would rather we didn't try to be smart here and just do things the dumb slow way to be safer.

Reply:
Question: I'm guessing we get the zeros added back to the end because "newTar" has them?

Reply:
Specifically this works with the tar files _we actually generate_.

Doesn't matter though, this is one of those situations where it'll probably take less time for me to immediately give up and rewrite it than for me to try and argue the case, so I'll just take the time to rewrite it

> Question: I'm guessing we get the zeros added back to the end because "newTar" has them?

the golang tar package follows the spec and writes them when you call `Close` so yeah newTar has the zeroes

Reply:
Actually, I sort-of changed my mind; rewriting this to end up with something that has exactly the same effect in the end isn't the best use of my time unless I _have_ to do it. This works totally fine; it looks a little magic but it works with the tar files we actually generate from the cert-manager.

If your comment is a "I wouldn't have done it this way" or "I'd rather not do it this way" then that's fair enough and it's acknowledged. If it's "it must not be done this way" then it obviously needs discussion (in practice, it'll be a rewrite because I'm not going to waste time talking about it). I'm gonna leave it as is for now since your comment is phrased as "I would rather"